test/unit/test_devices/test_aireos_device.py

```diff
@@ -538,6 +640,137 @@ def test_show_list(mock_enable, aireos_send_command_timing):
+):
+    mock_aime.side_effect = [False, False, True, True, False]
+    with pytest.raises(FileTransferError) as fte:
+        aireos_device.transfer_image_to_ap(aireos_boot_image)
+    assert len(mock_aime.mock_calls) == 5
+    mock_config.assert_has_calls([mock.call("ap image swap all")])
+    mock_wait.assert_not_called()
+    mock_bo.assert_not_called()
+    assert fte.value.message == f"Unable to set all APs to use {aireos_boot_image}"
+
+
+@mock.patch.object(AIREOSDevice, "config")
+@mock.patch.object(AIREOSDevice, "_wait_for_ap_image_download")
+@mock.patch.object(AIREOSDevice, "boot_options", new_callable=mock.PropertyMock)
+@mock.patch.object(AIREOSDevice, "_ap_images_match_expected")
+@mock.patch.object(AIREOSDevice, "ap_boot_options", new_callable=mock.PropertyMock)
+def test_transfer_image_to_ap_transfer_primary(
+    mock_abo, mock_aime, mock_bo, mock_wait, mock_config, aireos_device, aireos_mock_path, aireos_boot_image
+):
+    mock_bo.return_value = {"primary": aireos_boot_image, "backup": None}
+    mock_aime.side_effect = [False, False, False, False, True]
+    assert aireos_device.transfer_image_to_ap(aireos_boot_image) is True
+    assert len(mock_aime.mock_calls) == 5
+    mock_config.assert_has_calls([mock.call("ap image predownload primary all")])
+    mock_wait.assert_called()
+    mock_bo.assert_called_once()
+
+
+@mock.patch.object(AIREOSDevice, "config")
+@mock.patch.object(AIREOSDevice, "_wait_for_ap_image_download")
+@mock.patch.object(AIREOSDevice, "boot_options", new_callable=mock.PropertyMock)
+@mock.patch.object(AIREOSDevice, "_ap_images_match_expected")
+@mock.patch.object(AIREOSDevice, "ap_boot_options", new_callable=mock.PropertyMock)
+def test_transfer_image_to_ap_transfer_secondary(
+    mock_abo, mock_aime, mock_bo, mock_wait, mock_config, aireos_device, aireos_mock_path, aireos_boot_image
+):
+    mock_bo.return_value = {"primary": None, "backup": aireos_boot_image}
+    mock_aime.side_effect = [False, False, False, True, True]
+    assert aireos_device.transfer_image_to_ap(aireos_boot_image) is True
+    assert len(mock_aime.mock_calls) == 5
+    mock_config.assert_has_calls([mock.call("ap image predownload backup all"), mock.call("ap image swap all")])
+    mock_wait.assert_called()
+    mock_bo.assert_has_calls([mock.call(), mock.call()])
+
+
+@mock.patch.object(AIREOSDevice, "config")
+@mock.patch.object(AIREOSDevice, "_wait_for_ap_image_download")
+@mock.patch.object(AIREOSDevice, "boot_options", new_callable=mock.PropertyMock)
+@mock.patch.object(AIREOSDevice, "_ap_images_match_expected")
+@mock.patch.object(AIREOSDevice, "ap_boot_options", new_callable=mock.PropertyMock)
+def test_transfer_image_to_ap_transfer_secondary_fail(
+    mock_abo, mock_aime, mock_bo, mock_wait, mock_config, aireos_device, aireos_mock_path, aireos_boot_image
+):
+    mock_bo.return_value = {"primary": None, "backup": aireos_boot_image}
+    mock_aime.side_effect = [False, False, False, True, False]
```

Comment:
How about other combinations of these function return values, such as False, False, True, False, True? How about the first two arguments being true?

Reply:
This is a little tricky; if the 2nd response is True, then one of the responses is skipped (python returns early when the result of the conditional is determined, so True exists the `or` condition after the first test, and the second test never happens). The `[False, False, True, False, True]` is the same test as `[False, True, False, True]`, which is tested in `test_transfer_image_to_ap_already_transferred_primary`

[True, True] is not possible since the first true triggers the return value. This behavior is tested in `test_transfer_image_to_ap_already_active`